File: Examples/objective-c/Proximity/Proximity/Estimote/CloudManager.m
//
// Please report any problems with this app template to contact@estimote.com
//

#import "CloudManager.h"
#import <EstimoteSDK/EstimoteSDK.h>
#import "UIColor+Additions.h"

@implementation CloudManager

- (void)fetchColorsForBeaconsWithIdentifiers:(NSArray<NSString *> *)identifiers completion:(RequestCompletionBlock)completion {
    ESTRequestGetBeaconsDetails *request = [[ESTRequestGetBeaconsDetails alloc] initWithMacAddresses:identifiers andFields: ESTBeaconDetailsFieldColor];
    [request sendRequestWithCompletion:^(NSArray<ESTBeaconVO *> *beaconVOs, NSError *error) {
        if (error) {
            completion(false, nil);
            return;
        }
        
        NSMutableDictionary<NSString *, UIColor *> *colorByIdentifier = [NSMutableDictionary new];
        for (ESTBeaconVO *beaconVO in beaconVOs) {
            UIColor *color = [UIColor colorForEstimoteColor:beaconVO.color];
            colorByIdentifier[beaconVO.macAddress] = color;
        }
        
        completion(true, colorByIdentifier);
    }];
}

@end

File: Examples/swift/Proximity/Proximity/Estimote/CloudManager.swift
//
// Please report any problems with this app template to contact@estimote.com
//

import Foundation
import UIKit

/// Provides an API for fetching content related to beacons, in this simple case it's colors.
class CloudManager {
    
    enum RequestResult<T> {
        case success(T)
        case error
    }
    
    func fetchColors(beaconIdentifiers: [String], completion: @escaping ((RequestResult<[String: UIColor]>) -> ())) {
        
        let request = ESTRequestGetBeaconsDetails(macAddresses: beaconIdentifiers, andFields: ESTBeaconDetailsFields.fieldColor)
        
        request.sendRequest(completion: { (beaconInfos, error) in
            guard error == nil, let beaconInfos = beaconInfos as? [ESTBeaconVO] else {
                completion(.error)
                return
            }
            
            var colorByIdentifier = [String: UIColor]()
            for beaconInfo in beaconInfos {
                let color = UIColor.colorForEstimoteColor(beaconInfo.color)
                colorByIdentifier[beaconInfo.macAddress] = color
            }
            completion(.success(colorByIdentifier))
        })
    }
}
